Runtime Error encountered when using the Geneos utility to start a gateway

Background:

1. I installed a gateway in a client's UAT server using the Geneos utility.
2. a runtime error was encountered when I try to start the gateway I just created
3. Operating system of the UAT server is RHEL 7.9

Full Error Details:

$ geneos start gateway UAT_GATEWAY_TEST
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x89e357]

goroutine 1 [running]:
github.com/itrs-group/cordial/pkg/host.(*Local).Username(0xb468e0?)
        /app/cordial/pkg/host/local.go:51 +0x17
github.com/itrs-group/cordial/tools/geneos/internal/instance.Start({0xb468e0?, 0xc00007a0a0})
        /app/cordial/tools/geneos/internal/instance/start.go:42 +0xce
github.com/itrs-group/cordial/tools/geneos/cmd.Start.func1({0xb468e0?, 0xc00007a0a0?}, {0x1b?, 0xc0002078d0?, 0x97f660?})
        /app/cordial/tools/geneos/cmd/start.go:64 +0x25
github.com/itrs-group/cordial/tools/geneos/internal/instance.ForAll(0xc000012a08?, {0xa458c4, 0x3}, 0xaaaab0, {0xc000025300, 0x1, 0x4}, {0x                                                                                                  0, 0x0, 0x0})
        /app/cordial/tools/geneos/internal/instance/instance.go:465 +0x2f7
github.com/itrs-group/cordial/tools/geneos/cmd.Start(0xe70780?, 0x0, {0xc000025300, 0x1, 0x4}, {0x0, 0x0, 0x0})
        /app/cordial/tools/geneos/cmd/start.go:63 +0x90
github.com/itrs-group/cordial/tools/geneos/cmd.glob..func28(0xe70780?, {0xc00003a220?, 0x2?, 0x2?})
        /app/cordial/tools/geneos/cmd/start.go:58 +0x39
github.com/spf13/cobra.(*Command).execute(0xe70780, {0xc00003a1c0, 0x2, 0x2})
        [/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940](mailto:/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940) +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xe6f920)
        [/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068](mailto:/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068) +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        [/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992](mailto:/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992)
github.com/itrs-group/cordial/tools/geneos/cmd.Execute()
        /app/cordial/tools/geneos/cmd/root.go:231 +0x25
main.main()
        /app/cordial/tools/geneos/main.go:72 +0xfd

As part of troubleshooting, I tried to start the gateway manually and I was able to start it.

$ ./gateway2.linux_64 -setup /home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml -licd-host geneos-licd2.jpmchase.net -licd-port 6600
2023-06-23 02:52:58.371-0500 ***********************************************************
2023-06-23 02:52:58.371-0500 Geneos EMF2 Gateway GA6.3.1-230427 (c) 1995-2023 ITRS Group
2023-06-23 02:52:58.371-0500 ***********************************************************
2023-06-23 02:52:58.373-0500  INFO: Build type: linux_64:Release
2023-06-23 02:52:58.373-0500  INFO: Running from: /home/uat/ugeneos2/bin/geneos/packages/gateway/6.3.1
2023-06-23 02:52:58.373-0500  INFO: User/uid/gid: ugeneos2/92170/11536
2023-06-23 02:52:58.373-0500  INFO: Current path: /usr/lib64/qt-3.3/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/puppetlabs/bin:/home/uat/ugeneos2/bin/geneos
2023-06-23 02:52:58.373-0500  INFO: LD_LIBRARY_PATH has not been set
2023-06-23 02:52:58.373-0500  INFO: Gateway Gateway statistics collection is off.
2023-06-23 02:52:58.376-0500  INFO: SSLContext TLS version 1.0 and upwards accepted for secure connections.
2023-06-23 02:52:58.383-0500  INFO: CURL libcurl version: 7.69.1-DEV
2023-06-23 02:52:58.383-0500  INFO: CURL ssl version: OpenSSL/1.1.1q
2023-06-23 02:52:58.383-0500  INFO: CURL Support for: SSL NTLM IPv6 SPNEGO;  Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
2023-06-23 02:52:58.444-0500  INFO: LicenceManager Connecting to Licence Daemon ("***removing this info - Gian")
2023-06-23 02:52:58.444-0500  INFO: HostInfo: "***removing this info - Gian", Version #1 SMP Mon Dec 19 10:44:06 UTC 2022, Platform Linux, Release 3.10.0-1160.83.1.el7.x86_64, MachinS Linux 3.10.0-1160.83.1.el7.x86_64
2023-06-23 02:52:58.555-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Read from Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml=>5984ecf4f30e6d823c428a863
2023-06-23 02:52:58.556-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Read from Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml=>726e100d3d4e6482786667bee
2023-06-23 02:52:58.556-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml
2023-06-23 02:52:58.556-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml
2023-06-23 02:52:58.561-0500  INFO: SetupManager Applying setup.
2023-06-23 02:52:58.561-0500  INFO: Gateway Listening on INSECURE port 7039
2023-06-23 02:52:58.561-0500  INFO: Gateway Unique id is 18694698
2023-06-23 02:52:58.561-0500  INFO: Gateway Unique id was generated using the gateway name.
2023-06-23 02:52:58.561-0500  INFO: Gateway Gateway name is UAT_GATEWAY_TEST
2023-06-23 02:52:58.561-0500  INFO: Gateway history rollover file limit set to 10
2023-06-23 02:52:58.561-0500  INFO: Log archive script is not set.
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER sleepInTryHeartbeat is true
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER tryHeartbeatIterations is set to 10
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER heartbeatInterval is set to 75
2023-06-23 02:52:58.561-0500  INFO: CONMANAGER connectWait time is set to 30
2023-06-23 02:52:58.566-0500  INFO: SchemaManager The working version is now GA6.3.1-230427
2023-06-23 02:52:58.567-0500  INFO: SelfAnnouncingProbeManager Self-announcing probes feature now enabled. Probe limit is unlimited.
2023-06-23 02:52:58.568-0500  INFO: RuleManager Using 0 rule worker threads
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:6a19a77b8a1cac5e197e5aaf0f00fe2e | Write to Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/gateway.setup.xml=>5984ecf4f30e6d823c428a865
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Read from File : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml
2023-06-23 02:52:58.569-0500  INFO: SetupManager MD5Hash:be649c8620983b8e5eecffe2a3b798df | Write to Cache : File:/home/uat/ugeneos2/bin/geneos/gateway/gateways/UAT_GATEWAY_TEST/instance.setup.xml=>726e100d3d4e6482786667bde
2023-06-23 02:52:58.570-0500  INFO: SetupManager Finished applying setup.
2023-06-23 02:52:58.620-0500  WARN: GatewayLicenceManager Gateway is not licensed.
2023-06-23 02:52:58.670-0500  INFO: HOSTRESOLVER Timeout set to  15 seconds
2023-06-23 02:52:58.670-0500  INFO: HOSTRESOLVER resolver using  gethostbyname_r()
2023-06-23 02:52:58.670-0500  INFO: HOSTRESOLVER cache status is Enabled
2023-06-23 02:52:58.671-0500  INFO: HOSTRESOLVER operating mode is THREADED
2023-06-23 02:52:58.671-0500  INFO: HOSTRESOLVER trying to resolve "***removing this info - Gian"
2023-06-23 02:52:58.708-0500  INFO: HOSTRESOLVER Took 0 Seconds to resolve host ["***removing this info - Gian"]
2023-06-23 02:52:58.708-0500  INFO: HOSTRESOLVER host "***removing this info - Gian" resolved to "***removing this info - Gian"
2023-06-23 02:52:58.797-0500  INFO: HOSTRESOLVER Address info service: Max acceptable DNS lookup time 15 seconds
2023-06-23 02:52:58.797-0500  INFO: HOSTRESOLVER Duration DNS lookup suspension on failure, 300 seconds
2023-06-23 02:52:59.523-0500  INFO: LicenceManager Connected to Licensing Daemon
2023-06-23 02:52:59.823-0500  INFO: GatewaySystem Gateway is licenced.
2023-06-23 02:52:59.823-0500  INFO: RoleManager Gateway is now running as Stand Alone
2023-06-23 02:52:59.823-0500  INFO: RoleManager Gateway switching to Active state
2023-06-23 02:52:59.826-0500  INFO: RoleManager Gateway is now Active
2023-06-23 02:53:00.131-0500  INFO: SelfAnnouncingProbeManager Self Announcing Probes are licensed.
^C2023-06-23 02:54:16.151-0500  INFO: Gateway Signal INT received - terminating
2023-06-23 02:54:16.225-0500  INFO: Gateway Exiting.
2023-06-23 02:54:16.227-0500  INFO: DatabaseLogging Disabled

Has anyone experienced this error before? As an FYI, I performed the same work of installing the Geneos utility, gateway installation and started the gateway on the client's DEV server and it's working fine. That server is running the same RHEL 7.9 OS.

--- a/pkg/host/local.go
+++ b/pkg/host/local.go
@@ -47,8 +47,10 @@ func NewLocal() Host {
 }
 
 func (h *Local) Username() string {
-	u, _ := user.Current()
-	return u.Username
+	if u, err := user.Current(); err == nil {
+		return u.Username
+	}
+	return os.Getenv("USER")
 }
 
 func (h *Local) Hostname() string {

--- a/pkg/host/ssh.go
+++ b/pkg/host/ssh.go
@@ -78,8 +78,11 @@ type SSHOptions func(*SSHRemote)
 
 func evalOptions(r *SSHRemote, options ...any) {
 	// defaults
-	u, _ := user.Current()
-	r.username = u.Username
+	if u, err := user.Current(); err == nil {
+		r.username = u.Username
+	} else {
+		r.username = os.Getenv("USER")
+	}
 
 	for _, opt := range options {
 		switch o := opt.(type) {

--- a/pkg/process/process.go
+++ b/pkg/process/process.go
@@ -226,8 +226,11 @@ func Start(h host.Host, program Program, options ...Options) (pid int, err error
 		}
 	} else {
 		if program.Username == "" {
-			u, _ := user.Current()
-			program.Username = u.Username
+			if u, err := user.Current(); err == nil {
+				program.Username = u.Username
+			} else {
+				program.Username = os.Getenv("USER")
+			}
 		}
 
 		if program.WorkingDir == "" {

--- a/tools/geneos/cmd/deploy.go
+++ b/tools/geneos/cmd/deploy.go
@@ -149,15 +149,18 @@ var deployCmd = &cobra.Command{
 		if h == geneos.LOCAL {
 			if geneos.Root() == "" {
 				if deployCmdGeneosHome == "" {
-					var input string
-					u, _ := user.Current()
-					root := u.HomeDir
-					if filepath.Base(u.HomeDir) != Execname {
-						root = filepath.Join(u.HomeDir, Execname)
+					var input, home string
+					if u, err := user.Current(); err == nil {
+						home = u.HomeDir
+					} else {
+						home = os.Getenv("HOME")
 					}
-					input, err = config.ReadUserInput("Geneos Directory (default %q): ", root)
+					if filepath.Base(home) != Execname {
+						home = filepath.Join(home, Execname)
+					}
+					input, err = config.ReadUserInput("Geneos Directory (default %q): ", home)
 					if err == nil {
-						root = input
+						home = input
 					} else if err != config.ErrNotInteractive {
 						return
 					}

--- a/tools/geneos/cmd/hostcmd/add.go
+++ b/tools/geneos/cmd/hostcmd/add.go
@@ -26,6 +26,7 @@ import (
 	_ "embed"
 	"fmt"
 	"net/url"
+	"os"
 	"os/user"
 	"path/filepath"
 	"runtime"
@@ -111,8 +112,14 @@ geneos host add remote1 ssh://server.example.com/opt/geneos
 		log.Debug().Msgf("hostname: %s", sshurl.Hostname())
 
 		hostcf.SetDefault("hostname", sshurl.Hostname())
-		u, _ := user.Current()
-		hostcf.SetDefault("username", u.Username)
+		var username string
+		if u, err := user.Current(); err == nil {
+			username = u.Username
+		} else {
+			username = os.Getenv("USER")
+		}
+
+		hostcf.SetDefault("username", username)
 		hostcf.SetDefault("port", 22)
 		// XXX default to remote user's home dir, not local
 		hostcf.SetDefault(cmd.Execname, geneos.Root())

--- a/tools/geneos/cmd/initcmd/init.go
+++ b/tools/geneos/cmd/initcmd/init.go
@@ -171,7 +171,7 @@ var initTLSCmd = &cobra.Command{
 // initProcessArgs works through the parsed arguments and returns a
 // geneos.GeneosOptions slice to be passed to worker functions
 func initProcessArgs(args []string) (options []geneos.Options, err error) {
-	var homedir, root string
+	var root string
 
 	options = []geneos.Options{
 		geneos.Version(initCmdVersion),
@@ -186,12 +186,11 @@ func initProcessArgs(args []string) (options []geneos.Options, err error) {
 		}
 	}
 
-	u, err := user.Current()
-	homedir = "/"
-	if err != nil {
-		log.Error().Err(err).Msg("cannot get user details")
-	} else {
+	homedir := "/"
+	if u, err := user.Current(); err == nil {
 		homedir = u.HomeDir
+	} else {
+		homedir = os.Getenv("HOME")
 	}
 
 	log.Debug().Msgf("%d %v", len(args), args)

